test/test_framework.py

```diff
@@ -1369,3 +1377,144 @@ class StatefulObject(Object):
         self.assertEqual(parent._stored.bar, 4)
         # TODO: jam 2020-01-30 is there a clean way to tell that
         #       parent._stored._data.dirty is False?
+
+
+@patch('sys.stderr', new_callable=io.StringIO)
+class BreakpointTests(unittest.TestCase):
+
+    def setUp(self):
+        tmpdir = Path(tempfile.mkdtemp())
+        self.addCleanup(shutil.rmtree, str(tmpdir))
+
+        self.framework = Framework(tmpdir / "framework.data", tmpdir, None, None)
+        self.addCleanup(self.framework.close)
+
+    def test_ignored(self, fake_stderr):
+        # it doesn't do anything really unless proper environment is there
+        assert JUJU_DEBUG_ENVVAR not in os.environ
+
+        with patch('pdb.Pdb.set_trace') as mock:
+            self.framework.breakpoint()
+        self.assertEqual(mock.call_count, 0)
+        self.assertEqual(fake_stderr.getvalue(), "")
+
+    def test_pdb_properly_called(self, fake_stderr):
+        # the debugger needs to leave the user in the frame where the breakpoint is executed,
+        # which for the test is the frame we're calling it here in the test :)
+        with patch.dict(os.environ, {JUJU_DEBUG_ENVVAR: JUJU_DEBUG_ALL}):
+            with patch('pdb.Pdb.set_trace') as mock:
+                this_frame = inspect.currentframe()
+                self.framework.breakpoint()
+        self.assertEqual(mock.call_count, 1)
+        self.assertEqual(mock.call_args, ((this_frame,), {}))
```

Comment:
I'm not a fan of mocking and generally encourage people to do without, but this is a hard one to test the desired behavior outcome instead of the implementation of it, so sounds okay.

Reply:
Mocking stuff is a slippery slope, indeed. I use them for cases where "external" resources are used, and never using the "automatic attributes" mock magic, just a simple way to do patching and checking calls, maybe returning a specific value or raising an exception, etc.